lib/xmalloc.h

```diff
@@ -48,15 +48,15 @@
 /* for free() */
 #include <stdlib.h>
 
-extern void *xmalloc (size_t size);
-extern void *xzmalloc (size_t size);
-extern void *xcalloc (size_t nmemb, size_t size);
-extern void *xrealloc (void *ptr, size_t size);
-extern char *xstrdup (const char *str);
+__attribute__((malloc, warn_unused_result)) void *xmalloc (size_t size);
+__attribute__((malloc, warn_unused_result)) void *xzmalloc (size_t size);
+__attribute__((malloc, warn_unused_result)) void *xcalloc (size_t nmemb, size_t size);
+__attribute__((malloc, warn_unused_result)) void *xrealloc (void *ptr, size_t size);
```

Comment:
`xrealloc()` must not have the `malloc` attribute, according to gcc docs (emphasis mine):

> Using the attribute with no arguments is designed to improve optimization by relying on the aliasing property it implies. Functions like malloc and calloc have this property because they return a pointer to uninitialized or zeroed-out, newly obtained storage. **However, functions like realloc do not have this property, as they may return pointers to storage containing pointers to existing objects**. Additionally, since all such functions are assumed to return null only infrequently, callers can be optimized based on that assumption.

The other usages of `malloc`, and the other attributes, all look good.